### PR TITLE
Multicaster encapsulation

### LINKED ISSUE #743: `ConcurrentModificationException` when `unsubscribe` then `detach` channel presence listener
A customer is experiencing a crash with `ably-java` [version 1.2.10](https://github.com/ably/ably-java/releases/tag/v1.2.10), therefore my understanding is that this is pure JVM (i.e. a server type application - I think it's some kind of a bot), not Android. They have described the scenario to us:

> When the UserA encounters a new user join to chatroom he was in , it registers the user presence listener.

```java
presenceChannel.presence.subscribe(member -> {
```

> When UserA no longer needs to track the users presence (not a member of any active chat rooms), the listener is removed:

```java
presenceChannel.presence.unsubscribe();
try {
    presenceChannel.detach();
} catch (AblyException e) {
    throw new BotException(e.toString());
}
```

> The error is encountered mostly when a user goes offline....
> The following is the exception stack:

```
2022-01-25T15:37:29,129|INFO|SomeClient|WebSocketConnectReadThread-132|{"l":"DEBUG","msg":"receive presence event, action: leave, data: {\"userId\":\"SomeUserIdValue\",\"status\":\"Online\",\"customStatus\":\"Available From SomeNamedTool\"}","fnc":"SomeClient.registerUserPresence"} 
2022-01-25T15:37:29,131|DEBUG|UserPresenceManager|WebSocketConnectReadThread-132|updateUserPresence: userId=PATESTFICA138, source=MESSAGING, state=OFFLINE 
(ERROR): io.ably.lib.transport.WebSocketTransport: Unexpected exception processing received binary message 

io.ably.lib.types.AblyException: java.util.ConcurrentModificationException 
    at io.ably.lib.types.AblyException.fromThrowable(AblyException.java:54) 
    at io.ably.lib.transport.ConnectionManager.onMessage(ConnectionManager.java:1080) 
    at io.ably.lib.transport.WebSocketTransport$WsClient.onMessage(WebSocketTransport.java:161) 
    at org.java_websocket.client.WebSocketClient.onWebsocketMessage(WebSocketClient.java:505) 
    at org.java_websocket.drafts.Draft_6455.processFrameBinary(Draft_6455.java:835) 
    at org.java_websocket.drafts.Draft_6455.processFrame(Draft_6455.java:794) 
    at org.java_websocket.WebSocketImpl.decodeFrames(WebSocketImpl.java:381) 
    at org.java_websocket.WebSocketImpl.decode(WebSocketImpl.java:218) 
    at org.java_websocket.client.WebSocketClient.run(WebSocketClient.java:425) 
    at java.base/java.lang.Thread.run(Thread.java:829) 

Caused by: java.util.ConcurrentModificationException 
    at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1043) 
    at java.base/java.util.ArrayList$Itr.next(ArrayList.java:997) 
    at io.ably.lib.realtime.Presence$Multicaster.onPresenceMessage(Presence.java:359) 
    at io.ably.lib.realtime.Presence.broadcastPresence(Presence.java:345) 
    at io.ably.lib.realtime.Presence.setPresence(Presence.java:334) 
    at io.ably.lib.realtime.ChannelBase.onPresence(ChannelBase.java:769) 
    at io.ably.lib.realtime.ChannelBase.onChannelMessage(ChannelBase.java:1158) 
    at io.ably.lib.realtime.AblyRealtime$InternalChannels.onMessage(AblyRealtime.java:187) 
    at io.ably.lib.transport.ConnectionManager.onChannelMessage(ConnectionManager.java:1090) 
    at io.ably.lib.transport.ConnectionManager.onMessage(ConnectionManager.java:1075) 
    ... 8 more
```

(some details redacted above)

Hubspot (internal) threads conversing with this customer are [here](https://app.hubspot.com/live-messages/6939709/inbox/2125159045#email-reply-editor) and, then, [here](https://app.hubspot.com/live-messages/6939709/inbox/2126124055#email-reply-editor).

┆Issue is synchronized with this [Jira Uncategorised](https://ably.atlassian.net/browse/SDK-1524) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by QuintinWillison**
We've had a customer experiencing a `ConcurrentModificationException`, thrown by one of the `Multicaster` subclasses.

This isn't really that easy to test, without perhaps some sort of multi-threaded 'chaos' test... however, I think the problem was fairly obvious (as pointed out by @paddybyers in [this internal Slack message](https://ably-real-time.slack.com/archives/C8SPU4589/p1643285690094800?thread_ts=1643263851.076300&cid=C8SPU4589)), so the change presented in this pull request should make this base class robust to manipulation and query from multiple threads simultaneously.

I acknowledge that this does add a new shallow array copy operation each time an event requires member list iteration, but I've seen no evidence that this needed to be a high performance area of the codebase. We can further refactor in future if this is proven to be a performance hit.

Fixes #743.

### CODE REVIEW

**Review by ikbalkaya — APPROVED** [review]
As far as I can see this should fix the issue of synchronisation.
I would personally be inclined to use [synchronisedList](https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#synchronizedList(java.util.List)) for manipulation operations, however this still requires manual synchronisation when looping on the list, which was the issue in this case.

**Review by paddybyers — APPROVED** [review]
LGTM, thanks

**Review by KacperKluka — APPROVED** [review]
LGTM :+1:
